pcap-dpdk.c

```diff
@@ -0,0 +1,800 @@
+		}
+		// adjust rx tx
+		ret = rte_eth_dev_adjust_nb_rx_tx_desc(portid, &nb_rxd, &nb_txd);
+		if (ret < 0)
+		{
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: Cannot adjust number of descriptors: err=%d, port=%u",
+			    ret, portid);
+			ret = PCAP_ERROR;
+			break;	
+		}
+		// get MAC addr
+		rte_eth_macaddr_get(portid, &(pd->eth_addr));
+		eth_addr_str(&(pd->eth_addr), pd->mac_addr, DPDK_MAC_ADDR_SIZE-1);
+
+		// init one RX queue
+		rxq_conf = dev_info.default_rxconf;
+		rxq_conf.offloads = local_port_conf.rxmode.offloads;
+		ret = rte_eth_rx_queue_setup(portid, 0, nb_rxd,
+					     rte_eth_dev_socket_id(portid),
+					     &rxq_conf,
+					     pd->pktmbuf_pool);
+		if (ret < 0)
+		{
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: rte_eth_rx_queue_setup:err=%d, port=%u",
+			    ret, portid);
+			ret = PCAP_ERROR;
+			break;	
+		}
+
+		// init one TX queue 
+		txq_conf = dev_info.default_txconf;
+		txq_conf.offloads = local_port_conf.txmode.offloads;
+		ret = rte_eth_tx_queue_setup(portid, 0, nb_txd,
+				rte_eth_dev_socket_id(portid),
+				&txq_conf);
+		if (ret < 0)
+		{
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: rte_eth_tx_queue_setup:err=%d, port=%u",
+			    ret, portid);
+			ret = PCAP_ERROR;
+			break;	
+		}
+		// Initialize TX buffers 
+		tx_buffer = rte_zmalloc_socket(DPDK_TX_BUF_NAME,
+				RTE_ETH_TX_BUFFER_SIZE(MAX_PKT_BURST), 0,
+				rte_eth_dev_socket_id(portid));
+		if (tx_buffer == NULL)
+		{
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: Cannot allocate buffer for tx on port %u", portid);
+			ret = PCAP_ERROR;
+			break;	
+		}
+		rte_eth_tx_buffer_init(tx_buffer, MAX_PKT_BURST);
+		// Start device
+		ret = rte_eth_dev_start(portid);
+		if (ret < 0)
+		{
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: rte_eth_dev_start:err=%d, port=%u",
+			    ret, portid);
+			ret = PCAP_ERROR;
+			break;
+		}
+		// set promiscuous mode
+		if (p->opt.promisc){
+			pd->must_clear_promisc=1;
+			rte_eth_promiscuous_enable(portid);
+		}
+		// check link status
+		is_port_up = check_link_status(portid, &link);
+		if (!is_port_up){
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: link is down, port=%u",portid);
+			ret = PCAP_ERROR_IFACE_NOT_UP;
+			break;
+		}
+		// reset statistics
+		rte_eth_stats_reset(pd->portid);
+		calculate_timestamp(&(pd->ts_helper), &(pd->prev_ts));
+		rte_eth_stats_get(pd->portid,&(pd->prev_stats));	
+		// format pcap_t 
+		pd->portid = portid;
+		p->fd = pd->portid; 
+		if (p->snapshot <=0 || p->snapshot> MAXIMUM_SNAPLEN)
+		{
+			p->snapshot = MAXIMUM_SNAPLEN;
+		}
+		p->linktype = DLT_EN10MB; // Ethernet, the 10MB is historical.
+		p->selectable_fd = p->fd;
+		p->read_op = pcap_dpdk_dispatch;
+		p->inject_op = pcap_dpdk_inject;
+		// DPDK only support filter in userland now
```

Comment:
1. DPDK currently provides their own BPF API by rte_bpf.h. And it is also a user space solution.
https://doc.dpdk.org/guides/prog_guide/bpf_lib.html
2. However, DPDK will load BPF code from a ELF file by calling rte_bpf_elf_load(elf_file_name). It is difficult to use, because user needs compile BPF elf firstly, then pass it to DPDK programs.
https://www.dpdk.org/wp-content/uploads/sites/35/2018/10/pm-07-DPDK-BPFu6.pdf
3. Currently, I prefer use pcap_filter instead of rte_bpf.h. And later, if DPDK provides a user friendly API to compile BPF from string and make filtering more efficient, I will consider to change to rte_bpf.h
4. I will change the comment to make it clear.